ngl env (v2: configure.py+[n]make)

cboesch-gpsw opened the pull request.

jmoguillansky-gpsw commented:
Thanks for the improvement.  Some comments below

jmoguillansky-gpsw commented:
"Last but not least, the
added branches in the Makefile are becoming a maintainance nightmare."

>> Please remove this portion from the commit message (we already removed this issue in the makefile-based approach)

"Large part of the Windows support done by Matthieu Bouron."

>> Please reword this commit:
"Large part of the new design for Windows support done by Matthieu Bouron."
or something similar

jmoguillansky-gpsw commented:
"Pulling a whole Linux distribution
just for one tool (Make) is completely overkill."

>>Please remove this portion from the commit message.  That wasn't the only reason to use WSL.  It also provides pkg-config tool, bash script support, and a similar environment as Linux, with tools such as untar, etc.

jmoguillansky-gpsw commented:
I will integrate / test in multi-backend branch and get back to you shortly

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+
+@_block('tests-setup', [_ngl_tools_install, _pynodegl_utils_install])
+def _tests_setup(cfg):
+    return ['$(MESON_SETUP) ' + _cmd_join('tests', op.join('builddir', 'tests'))]
```
this won't work when backend is visual studio.  meson test only support ninja backend

cboesch-gpsw replied:
Fixed. Do you have a meson issue reference I can use here? There is no explanation on `master`.

jmoguillansky-gpsw replied:
it's not a meson bug, it's a meson design feature.  It will report an error if user chooses vs backend.

jmoguillansky-gpsw commented on .github/workflows/ci_msvc.yml:
```diff
@@ -42,17 +47,18 @@ jobs:
-          cd gopro-lib-node.gl
-          make -j$(($(nproc)+1)) TARGET_OS=Windows DEBUG_GPU_CAPTURE=yes \
-          VCPKG_DIR='C:\\vcpkg'
+          git clean -fdx  # trash the cached builds
```
why is this necessary on CI server?

cboesch-gpsw replied:
I thought it was because it fixed something during one of the iteration. I looked at the log and nothing is getting cleaned up so I probably got things mixed up here. Chunk removed.

jmoguillansky-gpsw commented on .github/workflows/ci_msvc.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
+        shell: powershell
 
     steps:
       - uses: Vampire/setup-wsl@v1
```
you can remove wsl

cboesch-gpsw replied:
Switched to `bash` as you suggested in the other PR.

jmoguillansky-gpsw commented on .github/workflows/ci_msvc.yml:
```diff
@@ -42,17 +47,18 @@ jobs:
+          git clean -fdx  # trash the cached builds
+          .\scripts\msvc-env.ps1
+          python.exe .\configure.py --debug gpu_capture
+          nmake
```
I think one thing missing from nmake is "dry run"?  Is there a way to enable?

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+    target = op.join(dst_base, name)
+    if _SYSTEM in ('Windows', 'MinGW'):
+        logging.info(f'copy {target} -> {extract_dir}')
+        shutil.copytree(extract_dir, target, dirs_exist_ok=True)
```
now we're not using wsl anymore.  Does os.symlink still not work on Windows?

cboesch-gpsw replied:
Yes it is necessary if you don't have the developer mode activated apparently.

jmoguillansky-gpsw replied:
can we check if it's supported and if so use symlink?
Python should return an "OSError" if not supported?  
We can store a flag somewhere if system supports symlinks?

https://docs.python.org/3/library/os.html
Note On newer versions of Windows 10, unprivileged accounts can create symlinks if Developer Mode is enabled. When Developer Mode is not available/enabled, the SeCreateSymbolicLinkPrivilege privilege is required, or the process must be run as an administrator.
OSError is raised when the function is called by an unprivileged user.

jmoguillansky-gpsw replied:
> > > "Pulling a whole Linux distribution
> > > just for one tool (Make) is completely overkill."
> > 
> > 
> > Can we reword this portion?  That wasn't the only reason to use WSL.  It also provides pkg-config tool, bash script support, and a similar environment as Linux, with tools such as untar, etc.  Maybe something like "Pulling a whole Linux distribution to support the build is completely overkill"
> 
> Sure, changed. Note that we're not relying on pkg-config nor bash script support currently afaict, but it's true for the external download thing.
> 
> > > "Last but not least, the
> > > added branches in the Makefile are becoming a maintainance nightmare."
> > 
> > 
> > Can we remove this portion from the commit message (we already removed this issue in the makefile-based approach)?
> 
> Sorry I didn't catch that: I'm referring to the `ifeq`/`else`/`endif` in the `Makefile` on master, which go away with this PR.

We have a fix already but it's not merged into master.  Anyhow it's up to you.  Maybe something like this would be better?
Last but not least, this removes the extra branches in the Makefile which are a maintanance nightmare

cboesch-gpsw replied:
Adjusted, thanks

cboesch-gpsw replied:
> can we check if it's supported and if so use symlink?
> Python should return an "OSError" if not supported?
> We can store a flag somewhere if system supports symlinks?

Done:
```python
    # Remove previous link if needed
    target = op.join(dst_base, name)
    rel_extract_dir = op.basename(extract_dir)
    if op.islink(target) and os.readlink(target) != rel_extract_dir:
        logging.info('unlink %s target', target)
        os.unlink(target)
    elif op.exists(target) and not op.islink(target):
        logging.info('remove previous %s copy', target)
        shutil.rmtree(target)

    # Link (or copy)
    if not op.exists(target):
        logging.info('symlink %s -> %s', target, rel_extract_dir)
        try:
            os.symlink(rel_extract_dir, target)
        except OSError:
            # This typically happens on Windows when Developer Mode is not
            # available/enabled
            logging.info('unable to symlink, fallback on copy (%s -> %s)', target, extract_dir)
            shutil.copytree(extract_dir, target)
```

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+
+@_block('pynodegl-deps-install', [_nodegl_install])
+def _pynodegl_deps_install(cfg):
+    return ['$(PIP) ' + _cmd_join('install', '-r', op.join('.', 'pynodegl', 'requirements.txt'))]
```
does this use system PIP or PIP in virtual environment?

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+                meson_setup += ['--libdir=lib']
+
+    return dict(
+        PIP=_cmd_join(python, '-m', 'pip'),
```
does this invoke system python or venv python?

cboesch-gpsw replied:
venv Python: `python = op.join(cfg.bin_path, 'python')`. I could rely on the `PATH` being set, but I wanted to explicit the path for this one in case `python` or `python3` is not in the venv and would fallback on the one in the system or that sort of stuff.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+
+    return dict(
+        PIP=_cmd_join(python, '-m', 'pip'),
+        MESON=meson,
```
does this invoke system meson or venv meson?

cboesch-gpsw replied:
venv meson unless not available (typically on MinGW), using the `PATH` being set.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,543 @@
+            vcpkg_bin = op.join(args.vcpkg_dir, 'installed', 'x64-windows', 'bin')
+            for f in glob.glob(op.join(vcpkg_bin, '*.dll')):
+                logging.info('copy %s to venv/Scripts', f)
+                shutil.copy2(f, op.join('venv', 'Scripts'))
```
I don't think this logic is in existing Makefile?
https://github.com/gopro/gopro-lib-node.gl/blob/master/Makefile

This is a bit dangerous: it copies all dlls from vcpkg folder?

cboesch-gpsw replied:
> I don't think this logic is in existing Makefile?
> https://github.com/gopro/gopro-lib-node.gl/blob/master/Makefile

It's not indeed, but we need to do that to be able to import properly the module until we have something better using add/set dll directory mechanism.

> This is a bit dangerous: it copies all dlls from vcpkg folder?

How dangerous exactly?

jmoguillansky-gpsw replied:
we just don't know what packages the user has, and what dlls are being copied, etc.  
It would be nice if we merge this commit also to avoid the DLL duplication, but we can do that later?
https://github.com/gopro/gopro-lib-node.gl/commit/e305182148705c5a8986ec60f673a53af5ec3de6

cboesch-gpsw replied:
Yeah but this PR is not ready; Matthieu has a prototype with a wrapper around `pynodegl` that does that, but it's not that trivial so it probably belongs in another PR.

jmoguillansky-gpsw replied:
I haven't updated the DLL PR, the link is to a specific commit.  I added an env variable NGL_DLL_DIRS and call add_dll_directory in the python module.  We can address it later.

cboesch-gpsw commented:
> > "Pulling a whole Linux distribution
> > just for one tool (Make) is completely overkill."
>
> Can we reword this portion?  That wasn't the only reason to use WSL.  It also provides pkg-config tool, bash script support, and a similar environment as Linux, with tools such as untar, etc.  Maybe something like "Pulling a whole Linux distribution to support the build is completely overkill"

Sure, changed. Note that we're not relying on pkg-config nor bash script support currently afaict, but it's true for the external download thing.

> > "Last but not least, the
> > added branches in the Makefile are becoming a maintainance nightmare."
> 
> Can we remove this portion from the commit message (we already removed this issue in the makefile-based approach)?

Sorry I didn't catch that: I'm referring to the `ifeq`/`else`/`endif` in the `Makefile` on master, which go away with this PR.

jmoguillansky-gpsw commented:
when I run configure from powershell, it spawns a new console window, and then the window disappears?
Is there a way to pipe the output to the powershell terminal directly?

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,553 @@
+        debug_opts = ','.join(cfg.args.debug)
+        nodegl_debug_opts += [f'-Ddebug_opts={debug_opts}']
+
+    if 'gpu_capture' in cfg.args.debug:
```
should this be inside if cfg.args.debug?  What is cfg.args.debug is None?

cboesch-gpsw replied:
I don't think that's possible to have `cfg.args.debug` set to `None` since the default  is explicitly set to `[]`

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,553 @@
+        self.bin_name = 'Scripts' if _SYSTEM == 'Windows' else 'bin'
+        self.bin_path = op.join(self.prefix, self.bin_name)
+        if _SYSTEM == 'Windows':
+            vcpkg_pc_path = op.join(args.vcpkg_dir, 'installed', 'x64-windows', 'lib', 'pkgconfig')
```
when using the argument parser, does it know to store this param to vcpkg_dir automatically?

 parser.add_argument('--vcpkg-dir'
 
 or do we need to add dest='vcpkg_dir'?

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,553 @@
+    return [f'copy {renderdoc_dll} {cfg.bin_path}']
+
+
+@_block('nodegl-setup', [_sxplayer_install])
```
how can I set the prerequisites for a target conditionally based on the command line args?

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,559 @@
+
+    if 'gpu_capture' in cfg.args.debug:
+        renderdoc_dir = cfg.externals['renderdoc']
+        nodegl_debug_opts += [f'-Drenderdoc_dir={renderdoc_dir}']
```
general suggestion: it might be better to make renderdoc_dir a makefile variable so that the generated makefile is a little cleaner, but it's not a requirement.  Similarly, for any other global variables in the makefile

cboesch-gpsw replied:
This is a bit annoying to do because we need a special escaping logic; `shlex.join` typically quotes `$(...)` while it shouldn't in the context of the `Makefile`. I could add heuristics in the quote mechanism to prevent that, but I'd rather have an ugly generated `Makefile` than a complex `configure.py`.

jmoguillansky-gpsw replied:
I see

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,559 @@
+def _fetch_externals(args):
+    dependencies = _EXTERNAL_DEPS.copy()
+    if _SYSTEM not in {'Windows', 'MinGW'}:
+        del dependencies['pkgconf']
```
maybe the code is a little simpler if we add the dependencies conditionally instead of removing them?

cboesch-gpsw replied:
> maybe the code is a little simpler if we add the dependencies conditionally instead of removing them

Yeah that wasn't great. I changed that to make it insert-based (and fixed a few things):
```python
def _get_external_deps(args):
    deps = ['sxplayer']
    if _SYSTEM == 'Windows':
        deps.append('pkgconf')
    if 'gpu_capture' in args.debug:
        if _SYSTEM not in {'Windows', 'Linux'}:
            raise Exception(f'Renderdoc is not supported on {_SYSTEM}')
        deps.append('renderdoc')
    return {dep: _EXTERNAL_DEPS[dep] for dep in deps}
```

> or when we define _EXTERNAL_DEPS, maybe we can have a field to specify which OS's?

A field is not enough because it's platform *and* options dependent; we'd need some lambda or dedicated functions to get availability.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,559 @@
+        'setup',
+        '--prefix', cfg.prefix,
+        '-Drpath=true',
+        '--pkg-config-path', cfg.pkg_config_path,
```
we can simplify generated makefile:
'--pkg-config-path', '$(PKG_CONFIG_PATH')

cboesch-gpsw replied:
Same issue as with renderdoc (quoting)

jmoguillansky-gpsw commented:
It seems like nmake invokes shell commands using command prompt, not using powershell?

jmoguillansky-gpsw commented:
"WSL (or MSYS2 when not using the MSVC toolchain) to make it work. "
>> You can use msys2 with msvc toolchain.
Please reword:
On Windows this means we need an overlay such as WSL or MSYS2

jmoguillansky-gpsw commented:
"Last but not least, this
removes the extra branches in the Makefile which are becoming a
maintainance nightmare."

>> Please remove this point.  We already submitted a PR 2 months ago which resolved this issue, so it's no longer relevant.

cboesch-gpsw commented:
> It seems like nmake invokes shell commands using command prompt, not using powershell?
> FYI: https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/nmake-fatal-error-u1056?view=msvc-160

I guess; is it an issue?

cboesch-gpsw commented:
> > "WSL (or MSYS2 when not using the MSVC toolchain) to make it work. "
>
> You can use msys2 with msvc toolchain.

Can we? I don't think this was ever supported in our build.

Please try to quote properly in markdown: `>>` is for double level quote, not for your answer.

jmoguillansky-gpsw commented:
> > It seems like nmake invokes shell commands using command prompt, not using powershell?
> > FYI: https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/nmake-fatal-error-u1056?view=msvc-160
> 
> I guess; is it an issue? How do you trigger that?

It's not an issue.  I just noticed when updating the configure script.

jmoguillansky-gpsw commented on .github/workflows/ci_msvc.yml:
```diff
@@ -18,19 +18,20 @@ jobs:
     defaults:
       run:
-        shell: wsl-bash {0}
+        shell: powershell
```
fyi: if you do shell: pwsh it will use newer version

cboesch-gpsw replied:
It appears to work, adjusted, thanks.

(For reference: https://stackoverflow.com/questions/60124810/what-is-the-difference-between-pwsh-and-powershell-integrated-console-on-vs)

cboesch-gpsw commented:
@jmoguillansky-gpsw did I address all your comments or is there something I missed?

jmoguillansky-gpsw commented:
> @jmoguillansky-gpsw did I address all your comments or is there something I missed?

Yes, I'm just finishing up integrating this into multi-backend branch, as soon as it's working i'll approve (should be today).
It would be nice if you can make some changes to the commit message but it's up to you:

On Windows this means we need an overlay such as WSL -> WSL or MSYS2
Last but not least, this removes the extra branches ... -> remove

jmoguillansky-gpsw commented on .github/workflows/ci_mac.yml:
```diff
@@ -20,5 +20,11 @@ jobs:
     - name: Build
       run: |
-        make -j$(($(sysctl -n hw.logicalcpu)+1))
+        ./configure.py
```
--buildtype debug --debug gl gpu_capture

jmoguillansky-gpsw commented on .github/workflows/ci_linux.yml:
```diff
@@ -37,8 +37,15 @@ jobs:
       run: |
-        make -j$(($(nproc)+1)) DEBUG_GPU_CAPTURE=yes
-    - name: Run tests ${{matrix.backend}}
+        ./configure.py --buildtype debug --debug gl gpu_capture
```
nit:
it might be better to pass debug args as comma-separated values, but up to you
parser.add_argument('--debug', type=str)
debug_list = parser.parse_args().debug.split(",")

cboesch-gpsw replied:
Renamed to `--debug-opts`. Not sure about the split, the arg parser has that feature builtin.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,564 @@
+    # Download
+    if not op.exists(dst_path):
+        logging.info('downloading %s to %s', url, dst_file)
+        urllib.request.urlretrieve(url, dst_path)
```
I started getting an SSL certificate error on the CI server.  Investigating...

I saw this API is deprecated, you may want to update it.  
Also may want to add some additional error handling

Traceback (most recent call last):
  File ".\configure.py", line 682, in <module>
    _run()
  File ".\configure.py", line 669, in _run
    cfg = _Config(args)
  File ".\configure.py", line 599, in __init__
    self.externals = _fetch_externals(args)
  File ".\configure.py", line 151, in _fetch_externals
    return dict(p.map(_download_extract, dependencies.items()))
  File "C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\multiprocessing\pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\multiprocessing\pool.py", line 771, in get
    raise self._value
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1125)>

jmoguillansky-gpsw replied:
seems like there's an issue with the pkgconf URL?
https://distfiles.dereferenced.org/pkgconf/pkgconf-@VERSION@.tar.xz

If you try to access in your browser: NET::ERR_CERT_DATE_INVALID

cboesch-gpsw replied:
It appears it's been fixed on their side.

cboesch-gpsw replied:
> I started getting an SSL certificate error on the CI server. Investigating...
> 
> I saw this API is deprecated, you may want to update it:
> 
> ```
> import requests
> ...
> response = requests.get(url)
> with open(dst_file, 'wb') as file:
> file.write(response.content)
> ```

Forgot to reply to that part:
- the API is not deprecated, it is just legacy ("They might become deprecated at some point in the future.")
- `requests` is an external dependency, we won't want to use that
- i looked at `urlretrieve` implementation and they're doing quite a bit of work which I'm not exactly willing to duplicate

I'll keep `urlretrieve` for now.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -0,0 +1,564 @@
+        dst_file='sxplayer-@VERSION@.tar.gz',
+    ),
+    pkgconf=dict(
+        version='1.7.3',
```
please use 1.7.4

cboesch-gpsw replied:
Bumped in an earlier commit.

Also bumped RenderDoc in a separate commit.

cboesch-gpsw commented:
> Last but not least, this removes the extra branches ... -> remove

Huh?

Anyway, I adjusted a few other bits here and there, notably:
- simplified ubuntu/debian checking (actually align it with the logic used in meson)
- added an external dependency checksum in case of mis-download (or if the configure is aborted during a download)

mbouron-gpsw commented:
Typo in the commit description: jungling -> juggling.

mbouron-gpsw commented on configure.py:
```diff
@@ -0,0 +1,581 @@
+    if _SYSTEM == 'Windows':
+        bindir = cfg.bin_path
+        ngl_tools_dlls = op.join('builddir', 'ngl-tools', '*.dll')
+        ret += [f'xcopy /Y {ngl_tools_dlls} {bindir}']
```
This seems unneeded on my machine, there is no dll anymore in this folder.

cboesch-gpsw replied:
removed

mbouron-gpsw commented:
I tested on my machine, linux, mingw, msvc and it works fine. Other than my remarks (and the other remaining ones), LGTM.

mbouron-gpsw commented on scripts/msvc-env.ps1:
```diff
@@ -9,4 +13,11 @@ $vsInstallationInfo = Get-VSSetupInstance -All -Prerelease |
+    }
+}
+
+Set-Item -force -Path env:VCVARS64 -Value "$vcvars64"
```
This line can be simplified to $env:VCVARS64 = "$vcvars64" but keeping the use of the Set-Item form makes the code consistent with what we do above. You can choose whatever you prefer.

mbouron-gpsw replied:
We can also use `Set-Item env:VCVARS64 "$vcvars64"` (to keep it consistent with the code above).

cboesch-gpsw replied:
applied your diff

mbouron-gpsw commented on scripts/msvc-env.ps1:
```diff
@@ -9,4 +13,11 @@ $vsInstallationInfo = Get-VSSetupInstance -All -Prerelease |
+cmd /c "call `"$vcvars64`" && set" |
+foreach {
+    if ($_ -match "=") {
+        $v = $_.split("="); Set-Item -force -Path "env:\$($v[0])" -Value "$($v[1])"
```
You can split this into two lines and remove `;`.

mbouron-gpsw replied:
`Set-Item -force -Path "env:\$($v[0])" -Value "$($v[1])"` can be simplified to `Set-Item -Force "env:$($v[0])" "$($v[1])"`
Note that the '\' before $() does not seem to do anything.

cboesch-gpsw replied:
applied your diff
